src/analyses/sese_regions.cpp

```diff
@@ -0,0 +1,261 @@
+  // we conservatively approximate the latter condition by checking that there
+  // are no loops with multiple backedges (this implies there are effectively
+  // two loops being described as one), and that there are no backedges not
+  // associated with a natural loop (indicating a cycle with multiple entry
+  // edges). Those conditions being met, it suffices to check that the would-be
+  // region has its entry and exit nodes in the same natural loops; without them
+  // we conservatively decline to identify any regions.
+
+  for(auto it = goto_program.instructions.begin();
+      it != goto_program.instructions.end();
+      ++it)
+  {
+    for(auto predecessor : it->incoming_edges)
+    {
+      if(
+        predecessor->location_number > it->location_number &&
+        !natural_loops.is_loop_header(it))
+      {
+        // No loop header associated with this backedge; conservatively decline
+        // to diagnose any SESE regions.
+        return;
+      }
+    }
+  }
+
+  for(const auto &loop : natural_loops.loop_map)
+  {
+    std::size_t backedge_count = std::count_if(
+      loop.first->incoming_edges.begin(),
+      loop.first->incoming_edges.end(),
+      [&](const goto_programt::const_targett &predecessor) {
+        return loop.first->location_number < predecessor->location_number;
+      });
+    if(backedge_count > 1)
+    {
+      // Loop with multiple backedges; conservatively decline to diagnose any
+      // SESE regions.
+      return;
+    }
+  }
+
+  // All checks passed, let's find some regions!
+  compute_sese_regions(goto_program, natural_loops);
+}
+
+typedef std::
+  unordered_map<goto_programt::const_targett, std::size_t, const_target_hash>
+    innermost_loop_mapt;
+
+/// Builds a map from instructions to the smallest (and therefore innermost)
+/// loop they are a member of.
+static innermost_loop_mapt
+compute_innermost_loop_ids(const natural_loopst &natural_loops)
+{
+  innermost_loop_mapt innermost_loop_ids;
+  std::vector<std::size_t> loop_size_by_id;
+
+  std::size_t loop_id = 0;
+  for(const auto &header_and_loop : natural_loops.loop_map)
+  {
+    const auto &loop = header_and_loop.second;
+    loop_size_by_id.push_back(loop.size());
+
+    for(const auto &instruction : loop)
+    {
+      auto emplace_result = innermost_loop_ids.emplace(instruction, loop_id);
+      if(!emplace_result.second)
+      {
+        // Is the existing loop for this instruction larger than this one? If so
+        // this is the inner of the two loops.
+        if(loop_size_by_id.at(emplace_result.first->second) > loop.size())
+          emplace_result.first->second = loop_id;
+      }
+    }
+
+    ++loop_id;
+  }
+
+  return innermost_loop_ids;
+}
+
+static long get_innermost_loop(
+  const innermost_loop_mapt &innermost_loops,
+  const goto_programt::const_targett &target)
+{
+  auto findit = innermost_loops.find(target);
+  if(findit == innermost_loops.end())
+    return -1;
```

Comment:
`optionalt`?

Reply:
I couldn't compile with `optionalt` due to gcc-5 bugs :(